examples/hdf5.py

```diff
@@ -4,99 +4,199 @@
     def __init__(self, *args, **kwds):
-        self.hdf5 = None
-        self.limit = 10000 # maximum number of samples to be plotted
+        self.sampler = None
         pg.PlotCurveItem.__init__(self, *args, **kwds)
         
-    def setHDF5(self, data):
-        self.hdf5 = data
-        self.updateHDF5Plot()
-        
+    def setDataSampler(self, sampler):
+        self.sampler = sampler
+        self.loadData()
+
     def viewRangeChanged(self):
-        self.updateHDF5Plot()
-        
-    def updateHDF5Plot(self):
-        if self.hdf5 is None:
+        self.loadData(lazy=True)
+
+    def loadData(self, lazy=False):
+        if self.sampler is None:
             self.setData([])
             return
-        
+
         vb = self.getViewBox()
         if vb is None:
             return  # no ViewBox yet
-        
-        # Determine what data range must be read from HDF5
-        xrange = vb.viewRange()[0]
-        start = max(0,int(xrange[0])-1)
-        stop = min(len(self.hdf5), int(xrange[1]+2))
-        
-        # Decide by how much we should downsample 
-        ds = int((stop-start) / self.limit) + 1
-        
-        if ds == 1:
-            # Small enough to display with no intervention.
-            visible = self.hdf5[start:stop]
-            scale = 1
-        else:
-            # Here convert data into a down-sampled array suitable for visualizing.
-            # Must do this piecewise to limit memory usage.        
-            samples = 1 + ((stop-start) // ds)
-            visible = np.zeros(samples*2, dtype=self.hdf5.dtype)
-            sourcePtr = start
-            targetPtr = 0
-            
-            # read data in chunks of ~1M samples
-            chunkSize = (1000000//ds) * ds
-            while sourcePtr < stop-1: 
-                chunk = self.hdf5[sourcePtr:min(stop,sourcePtr+chunkSize)]
-                sourcePtr += len(chunk)
-                
-                # reshape chunk to be integral multiple of ds
-                chunk = chunk[:(len(chunk)//ds) * ds].reshape(len(chunk)//ds, ds)
-                
-                # compute max and min
-                chunkMax = chunk.max(axis=1)
-                chunkMin = chunk.min(axis=1)
-                
-                # interleave min and max into plot data to preserve envelope shape
-                visible[targetPtr:targetPtr+chunk.shape[0]*2:2] = chunkMin
-                visible[1+targetPtr:1+targetPtr+chunk.shape[0]*2:2] = chunkMax
-                targetPtr += chunk.shape[0]*2
-            
-            visible = visible[:targetPtr]
-            scale = ds * 0.5
-            
-        self.setData(visible) # update the plot
-        self.setPos(start, 0) # shift to match starting index
-        self.resetTransform()
-        self.scale(scale, 1)  # scale to match downsampling
 
-        
+        x1, x2 = vb.viewRange()[0]
+        if not lazy or not self.sampler.isRangeValid(x1, x2):
+            x, y = self.sampler[x1:x2]
```

Comment:
this is your "lone method call" afaict.
Imo `.sampler` does not need to be a class.

Reply:
There's `isRangeValid` and `__getitem__`.  In addition to the cached data, the class also stores the last view range so it can determine if it has been invalidated.

Reply:
> In addition to the cached data, the class also stores the last view range so it can determine if it has been invalidated.

Yeah maybe it's this part I'm not understanding.

If you have a range that's been invalidated don't you have to call all the same methods again?

If so, why not just have the caller of this `cache_x_range()` func do the range checking and re-call the downsampling steps as needed from client code instead of assigning class vars that are only written once afaict and then having it all triggered by multiple `__getitem__()` calls?